eval.go

```diff
@@ -23,9 +23,68 @@ type EvalConfig struct {
 // semantic check on an AST tree. This will be called with the root node.
 type SemanticChecker func(ast.Node) error
 
+// EvalType represents the type of the output returned from a HIL
+// evaluation.
+type EvalType uint32
+
+const (
+	TypeInvalid EvalType = 0
+	TypeString  EvalType = 1 << iota
+	TypeList
+	TypeMap
+)
+
+//go:generate stringer -type=EvalType
+
+// EvaluationResult is a struct returned from the hil.Eval function,
+// representing the result of an interpolation. Results are returned in their
+// "natural" Go structure rather than in terms of the HIL AST.  For the types
+// currently implemented, this means that the Value field can be interpreted as
+// the following Go types:
+//     TypeInvalid: undefined
+//     TypeString:  string
+//     TypeList:    []interface{}
+//     TypeMap:     map[string]interface{}
+type EvaluationResult struct {
+	Type  EvalType
+	Value interface{}
+}
+
+// InvalidResult is a structure representing the result of a HIL interpolation
+// which has invalid syntax, missing variables, or some other type of error.
+// The error is described out of band in the accompanying error return value.
+var InvalidResult = EvaluationResult{Type: TypeInvalid, Value: nil}
+
+func Eval(root ast.Node, config *EvalConfig) (EvaluationResult, error) {
+	output, outputType, err := internalEval(root, config)
+	if err != nil {
+		return InvalidResult, err
+	}
+
+	switch outputType {
+	case ast.TypeList:
+		return EvaluationResult{
+			Type:  TypeList,
+			Value: hilListToGoSlice(output.([]ast.Variable)),
+		}, nil
+	case ast.TypeMap:
+		return EvaluationResult{
+			Type:  TypeMap,
+			Value: hilMapToGoMap(output.(map[string]ast.Variable)),
+		}, nil
+	case ast.TypeString:
+		return EvaluationResult{
+			Type:  TypeString,
+			Value: output,
+		}, nil
+	default:
+		return InvalidResult, fmt.Errorf("unknown type %s as interpolation output", outputType)
+	}
+}
+
 // Eval evaluates the given AST tree and returns its output value, the type
 // of the output, and any error that occurred.
-func Eval(root ast.Node, config *EvalConfig) (interface{}, ast.Type, error) {
+func internalEval(root ast.Node, config *EvalConfig) (interface{}, ast.Type, error) {
```

Comment:
Nice, this was a clean wrap around the old outer `Eval` func. That's easy/clean.